README


This archive contains the distribution Text-CSV-Merge,
version 0.04:

  Fill in gaps in a CSV file from another CSV file

This software is copyright (c) 2013 by Michael Gatto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




CHANGES
==================================================
Changes from 2012-09-11 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.04 at 2013-09-11 21:48:13 +0000
-----------------------------------------

  Change: cc33583029dc69cee084f23ddfcb0ecffd8da77e
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 14:47:05 +0000

    * Need it for dzil and its manifest... 

  Change: ef69d3b728e45be5f5e4b68d01cde26f0ef835d9
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 14:22:37 +0000

    Merge tag 'v0.04' into develop

    Release 0.04 v0.04 

-----------------------------------------
version 0.04 at 2013-09-11 21:21:21 +0000
-----------------------------------------

  Change: 07a6c46859a7496e4cf6e6580b88352806faa4d6
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 14:21:21 +0000

    Merge branch 'release/0.04' 

  Change: 7f04cf42f517f72664195d9488be723184ebc3fc
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 14:19:15 +0000

    * Minor formatting. 

  Change: 6060dcb623768498e1b130d87582e2d84cd40b0d
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 12:45:15 +0000

    * Fix columns definition in test. 

  Change: dfc06cf542c29ee4dc34ad73c00bb10e158a78c2
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 08:16:33 +0000

    * Fix test failures under Perl >= 5.18.

    The smart match operator is now experimental and produces warnings. I
    replaced it with grep(). 

  Change: f0e53706ee827ea3e5b1ea54feeb2074916c6ca8
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-09-11 07:59:20 +0000

    * Test result CSV file should not be in source code? 

  Change: 04bc71b1031924aceef3b4c36b6d876520120582
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 10:14:30 +0000

    * Fix overwriting of README.md with POD from main module. 

  Change: 5e9ff77cdb89c70f209b217912222b28a056afb0
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 10:12:45 +0000

    * Reverted change in which POD overwrote my github README.md.

    This is probably a result of the new build process, which I thought
    would only replace the generic README in the dist with this one... 

  Change: eab93fbaab26a5ea932d6e8b2fdb44845185f758
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 10:06:25 +0000

    Merge tag 'v0.03' into develop

    Prepared code for releasing a new, public version. 

-----------------------------------------
version 0.03 at 2013-08-26 17:05:37 +0000
-----------------------------------------

  Change: eb37498599ef1ce923334c31e9dfa1873a6245af
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 10:05:37 +0000

    Merge branch 'release/0.03' 

  Change: f7260543ff9a2e57cb84fed16654e344a3bf6d6c
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 08:14:05 +0000

    * Improve documentation.

    Fixed mistakes in POD formatting. Added docs for CLI utility. 

  Change: 562ba478cddff238f7d14ff95e25c0bffa874e93
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-08-26 08:12:04 +0000

    * Improved build process!

    Most things seem to work better with no missing items from Metacpan
    nor build errors nor oddities. 

  Change: 6e1e93dfcae91cfd9c4a1704157dcbe85a42397b
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-25 09:13:50 +0000

    * Fix POD headers. 

  Change: bf714fcd15f2f5660edf934a864e933bdf131502
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 21:52:38 +0000

    Bump version to 0.03 

  Change: 6827602463d5778e45a8dc5854db21f1b188aded
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 21:47:36 +0000

    * Move search field validation out of CLI and into lib. 

  Change: 3a292b255b07fabd61ab8e5d8d986b08a6659fb9
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 18:40:20 +0000

    Merge tag 'v0.002' into develop

    v0.002 

------------------------------------------
version 0.002 at 2013-07-25 01:37:58 +0000
------------------------------------------

  Change: 74e79571fb5a05c2c22a5af2f9bcfba23bf5a740
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 18:37:58 +0000

    Merge branch 'release/0.002' 

  Change: 7e13ef73be8c857012608e486ee68b83f4613a10
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 18:36:29 +0000

    * Finalize the build process. 

  Change: 113901164900a11182bdd811815fed4727ec79ea
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 09:30:20 +0000

    * Fix output file name for the MANIFEST. 

  Change: a422d320ea58de89746c3abc6f6b7bddbc29200d
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 09:28:22 +0000

    * Yet another test fix. 

  Change: f898c977cae75d7e102226bf0e6d095342501358
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 09:24:59 +0000

    * Fix next round of failed tests. 

  Change: 6090407ad165242a7d8c2ee617249a3f4d68056b
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 09:21:35 +0000

    * Fix errors in tests. 

  Change: 49fda6c7f8956db035383ee327d6d0be35b51001
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 09:09:25 +0000

    * Add ABSTRACT for Dist::Zilla. 

  Change: 7449c941c067ae059626491c9c51841c2d298092
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 08:40:29 +0000

    Merge branch 'master' into develop 

  Change: 8e80441689ce91ecd3b76c87d4812b85c90f945a
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 08:39:48 +0000

    * Test more extensively.

    Alert: some tests are not complete and do not conform to the intended
    test plan setting, yet. 

  Change: ff8233e4edcc15f6346b34a648efce3ead577817
  Author: Michael Gatto <mgatto@lisantra.com>
  Date : 2013-07-24 08:38:19 +0000

    * Add installation instructions. 

  Change: 6113083dba97d8a33e342491c6a76f51e7cd04be

lib/Text/CSV/Merge.pm
package Text::CSV::Merge;
{
  $Text::CSV::Merge::VERSION = '0.04';
}
# ABSTRACT: Fill in gaps in a CSV file from another CSV file

use Modern::Perl '2010';
use Moo 1.001000;
use IO::File;
use Text::CSV_XS;
use DBI; # for DBD::CSV
use Log::Dispatch;
use autodie;
use utf8;


has +logger => (
    is => 'lazy',
    builder => sub {
        Log::Dispatch->new(
            outputs => [
                [ 'File', autoflush => 1, min_level => 'debug', filename => 'merge.log', newline => 1, mode => '>>' ],
                #[ 'Screen', min_level => 'info', newline => 1 ],
            ],
        );        
    }
);

has +csv_parser => (
    is => 'lazy',
    builder => sub {
        Text::CSV_XS->new({ binary => 1, eol => $/ })
            or die "Cannot use CSV: " . Text::CSV_XS->error_diag();
    }
);

has +dbh => (
    is => 'lazy',
    # must return only a code ref
    builder => sub {    
        DBI->connect("dbi:CSV:", undef, undef, { 
            RaiseError => 1, 
            PrintError => 1, 
            f_ext => ".csv/r", 
            # Better performance with XS
            csv_class => "Text::CSV_XS", 
            # csv_null => 1, 
        }) or die "Cannot connect: $DBI::errstr";
    }
);

has base_file => (
    is => 'rw',
    required => 1,
    #allow external names to be different from class attribute
    init_arg => 'base',
    #validate it
    #isa => sub {},
    coerce => sub {
        my $base_fh = IO::File->new( $_[0], '<' ) or die "$_[0]: $!";
        $base_fh->binmode(":utf8");
        
        return $base_fh;
    }
);

has merge_file => (
    # We use only the raw file name/path and do not create a FH here, unlike base_file().
    is => 'rw',
    init_arg => 'merge',
    required => 1
);

has output_file => (
    is => 'rw',
    init_arg => 'output',
    # an output file name is NOT required
    required => 0,
    default => 'merged_output.csv',
    coerce => sub {
        my $output_fh = IO::File->new( $_[0], '>' ) or die "$_[0]: $!";
        $output_fh->binmode(":utf8");
        
        return $output_fh;
    }
);

has columns=> (
    is => 'rw',
    required => 1,
);    

has search_field => (
    is => 'rw',
    required => 1,
    init_arg => 'search'#,
    #isa => sub {
        # validate that search_field is one of the columns in the base file
        #die "Search parameter: '$_[0]' is not one of the columns: @{$self->columns}";
        #    unless ( $_[0] ~~ @{$self->columns} );
    #}
);

has first_row_is_headers => (
    is => 'rw',
    required => 1,
    #validate it
    isa => sub {
        # @TODO: there's got to be a better way to do this!
        die "Must be 1 or 0" unless ( $_[0] =~ /'1'|'0'/ || $_[0] == 1 || $_[0] == 0 );
    },
);

#=method BUILD
#Constructor. 
#=cut
#sub BUILD {
#    my $self = shift;
#}

sub merge {
    my $self = shift;

    # validate that search_field is one of the columns in the base file
    die "Search parameter: '$self->search_field' is not one of the columns: @{$self->columns}"
        unless ( scalar(grep { $_ eq $self->search_field } @{$self->columns}) );
        # Use scalar() to force grep to return the number of matches; 
        # 0 -> false for the 'unless' statement.
    
    $self->csv_parser->column_names( $self->columns );
        
    # Loop through the base file to find missing data
    #@TODO: make into $self->rows?
    my @rows;
    
    while ( my $row = $self->csv_parser->getline_hr( $self->base_file ) ) {
        # skip column names
        next if ($. == 1 and $self->first_row_is_headers);

        if ( $self->csv_parser->parse($row) ) {
            # keep a list of null columns in this row
            my @nulls;

            # might be slightly more efficient to use while()
            foreach my $key ( keys %{$row} ) {
                # which fields is this row missing?
                if ( $row->{$key} eq 'NULL' or $row->{$key} eq "" ) {
                    push @nulls, $key;

                    $self->logger->info("Missing data in column: $key for '$row->{$self->search_field}'");
                }
            }

            # make a hash of arrays
            if ( @nulls  ) {
                # search $merge_file for the missing data's row
                $" = ','; # reset the list separator for array interpolation to suit SQL
                
                # To get the original case for the columns, specify the column
                # names rather than using SELECT *, since it normalizes to
                # lowercase, per:
                # http://stackoverflow.com/questions/3350775/dbdcsv-returns-header-in-lower-case
                my $sth = $self->dbh->prepare(
                    "select @{$self->columns} from $self->{merge_file} where $self->{search_field} = ?"
                ) or die "Cannot prepare: " . $self->dbh->errstr ();

                $sth->execute($row->{$self->search_field});
                
                while ( my $filler = $sth->fetchrow_hashref() ) {
                    foreach my $gap ( @nulls ) {
                        if (exists $filler->{$gap} and defined $filler->{$gap} and $filler->{$gap} ne "") {
                            $self->logger->info(
                                "Found Data: '$gap' = '$filler->{$gap}' for '$row->{$self->search_field}'"
                            );
                            
                            $row->{$gap} = $filler->{$gap};
                        } else {
                            $self->logger->info(
                                "Data not Found for column: '$gap' for '$row->{$self->search_field}' $self->{merge_file}"
                            );
                        }
                    }
                }        
                
                # Be efficient and neat!
                $sth->finish();
            }
            
            # insert the updated row as a reference; even if not processed, the 
            # row will still appear in the final output.
            push @rows, $row;
        } else {
            my $err = $self->csv_parser->error_input;
            $self->logger->error("Failed to parse line: $err");
        }
    }

    # Ensure we've processed to the end of the file
    $self->csv_parser->eof or $self->csv_parser->error_diag();

    # print does NOT want an actual array! Use a hash slice, instead:
    #$self->csv_parser->print($output_fh, [ @$_{@columns} ]) for @rows;
    #
    # Or, here I've switched to Text::CSV_XS's specific print_hr(), which 
    # is simply missing from the PP (Pure Perl) version.
    $self->csv_parser->print_hr($self->output_file, $_) for @rows;
};

sub DEMOLISH {
    my $self = shift;

    ## Clean up!
    $self->base_file->close();
    $self->output_file->close() or die "output.csv: $!";
}


1;

__END__

=pod

=head1 NAME

Text::CSV::Merge - Fill in gaps in a CSV file from another CSV file

=head1 VERSION

version 0.04

=head1 SYNOPSIS

    my $merger = Text::CSV::Merge->new({
        base    => 'into.csv',
        merge   => 'from.csv',
        output  => 'output.csv',  # optional
        columns => [q/email name age/],
        search  => 'email',
        first_row_is_headers => 1  # optional
    });

    $merger->merge();
    
    ## Now, there is a new CSV file named 'merged_output.csv' by default, 
    #  in the same directory as the code which called C<$merger->merge();>.

=head1 DESCRIPTION

The use case for this module is when one has two CSV files of largely the same structure. Yet, the 'from.csv' has data which 'into.csv' lacks. 

In this initial release, Text::CSV::Merge only fills in empty cells; it does not overwrite data in 'into.csv' which also exists in 'from.csv'. 

=head2 Subclassing

Text::CSV::Merge may be subclassed. In the subclass, the following attributes may be overridden:

=over 4

=item *

C<csv_parser>

=item *

C<dbh>

=item *

C<logger>

=back

=head1 ATTRIBUTES

=head2 C<logger>

The logger for all operations in this module.

The logger records data gaps in the base CSV file, and records which data from the merge CSV file was used fill the gaps in the base CSV file.

=head2 C<csv_parser>

The CSV parser used internally is an immutable class property. 

The internal CSV parser is the XS version of Text::CSV: Text::CSV_XS. You may use Text::CSV::PP if you wish, but using any other parser which does not duplicate Text::CSV's API will probably not work without modifying the source of this module.

Text::CSV_XS is also used, hard-coded, as the parser for DBD::CSV. This is configurable, however, and may be made configurable by the end-user in a future release. It can be overridden in a subclass. 

=head2 C<dbh>

Create reusable DBI connection to the CSV data to be merged in to base file. 

This method is overridable in a subclass. A good use of this would be to merge data into an existing CSV file from a database, or XML file. It must conform to the DBI's API, however.

DBD::CSV is a base requirement for this module.

=head2 C<base_file>

The CSV file into which new data will be merged.

The base file is readonly, not read-write. This prevents accidental trashing of the original data.

=head2 C<merge_file>

The CSV file used to find data to merge into C<base_file>.

=head2 C<output_file>

The output file into which the merge results are written. 

I felt it imperative not to alter the original data files. I may make this a configurable option in the future, but wold likely set its default to 'false'.

=head2 C<columns>

The columns to be merged.

A column to be merged must exist in both C<base_file> and C<merge_file>. Other than that requirement, each file may have other columns which do not exist in the other.

=head2 C<search_field>

The column/field to match rows in C<merge_file>. 

This column must exist in both files and be identically cased.

=head2 C<first_row_is_headers>

1 if the CSV files' first row are its headers; 0 if not. 

If there are no headers, then the column names supplied by the C<columns> argument/property are applied to the columns in each file virtually, in numerical orders as they were passed in the list.

=head1 METHODS

=head2 C<merge()>

Main method and is public.

C<merge()> performs the actual merge of the two CSV files.

=head2 C<DEMOLISH()>

This method locally overrides a Moo built-in. 

It close out all file handles, which will only occur after a call to C<merge()>.

=head1 SEE ALSO

=over 4

=item *

L<Text::CSV_XS>

=back

=head1 AUTHOR

Michael Gatto <mgatto@lisantra.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Michael Gatto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Fill in gaps in a CSV file from another CSV file\r",
  "AUTHOR" => "Michael Gatto <mgatto\@lisantra.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Text-CSV-Merge",
  "EXE_FILES" => [
    "bin/merge_csv_files.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Text::CSV::Merge",
  "PREREQ_PM" => {
    "DBD::CSV" => 0,
    "DBI" => 0,
    "Getopt::Long" => 0,
    "IO::File" => 0,
    "Log::Dispatch" => 0,
    "Modern::Perl" => 0,
    "Moo" => "1.001000",
    "Text::CSV_XS" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



